### PR TITLE
ENH: add copy parameter for api.reshape function

### PR DESCRIPTION
**Pull request by bwalshe**
This adds a parameter to api.reshape to specify if data should be copied. This parameter is required so that api.reshape conforms to the standard. See #23410

### CODE REVIEW

File: numpy/array_api/_manipulation_functions.py
Comment (asmeurer) [inline numpy/array_api/_manipulation_functions.py:67]: I think this can be implemented by calling `reshape` and raising an exception if the result is not a view. Or maybe there's even a more direct way to tell if `reshape` will copy without actually trying it.

Comment (asmeurer) [reply numpy/array_api/_manipulation_functions.py:67]: I guess we need to implement it directly in `np.reshape`. Do others have issues with the above simple workaround here for now though?

Comment (bwalshe) [reply numpy/array_api/_manipulation_functions.py:67]: > I guess we need to implement it directly in np.reshape.

I think that in order to do this correctly, I might have to change the signature of the `PyArray_Newshape(PyArrayObject *self, PyArray_Dims *newdims, NPY_ORDER order)` function in `numpy/core/src/multiarray/shape.c`, but this has been marked as `NUMPY_API` in the comments.

I'll make the change you suggested above, but if I want to go ahead with the change to `PyArray_Newshape`, what do I have to do to get approval for that?

Comment (bwalshe) [reply numpy/array_api/_manipulation_functions.py]: @asmeurer Do these changes look OK?

**Comment by asmeurer** [comment]
This looks fine. Like I said, ideally we'd be able to detect if a copy will happen without actually doing it. Presumably this is just some function of the array strides. But since this also needs to be implemented in the main namespace eventually anyway, it would need to be implemented in the actual `np.reshape` function. So we can just use what is here for now. The minor performance hit isn't a huge concern for numpy.array_api because it's mainly used for testing.

**Comment by bwalshe** [comment]
I can emulate the check that happens inside `PyArray_Newshape`, but the issue I have with that is that if I re-implement it here and then the original changes, I think it's likely it will be missed that this needs to change as well. I could change `PyArray_Newshape` so that the check is some kind of independent function that could be used here as well, but that might be too big a change.

Which would you prefer?

**Comment by bwalshe** [comment]
@charris Apologies, but I have never made a PR to this project before - what does the `Needs Release Note` tag mean? Is this something I have to take action on?

**Comment by asmeurer** [comment]
> I can emulate the check that happens inside PyArray_Newshape, but the issue I have with that is that if I re-implement it here and then the original changes, I think it's likely it will be missed that this needs to change as well. 

I'm not sure if that would happen. The logic is based on whether a given shape is possible to represent with strides or not. At any rate, if this did happen it would be caught by the array-api-tests test suite. 

But like I said, I think what is here is fine for now. We are going to upstream this to `np.reshape` itself, in which case `np.array_api.reshape` will just become a thin wrapper around `np.reshape` again.

**Comment by bwalshe** [comment]
OK, I think I misunderstood your last message so. I thought you were asking me to make further changes. 

If this PR looks OK now, what do I do next?

**Comment by asmeurer** [comment]
I'm not actually a NumPy maintainer so someone else will need to merge this. My understanding is you do need a release note entry (see #5 here https://numpy.org/doc/stable/dev/index.html#development-process-summary)

**Comment by charris** [comment]
> what does the Needs Release Note tag mean?

They are small files containing short descriptions of what the PR does. Look in `doc/release/upcoming_changes` for examples. The the filename for this PR would be `23789.new_feature.rst`

**Comment by seberg** [comment]
Sorry, does anyone mind if I just remove the release note again or reduce it to a single bullet point?
I am not a fan of having a long release note snippet for a namespace that doesn't target end-user (and may well be removed again).

**Comment by bwalshe** [comment]
> Sorry, does anyone mind if I just remove the release note again or reduce it to a single bullet point? I am not a fan of having a long release note snippet for a namespace that doesn't target end-user (and may well be removed again).

NP. I removed everything after the first line. If you want to make other changes, go for it.

**Review by asmeurer — APPROVED** [review]

File: doc/release/upcoming_changes/23789.new_feature.rst
Comment (mattip) [inline doc/release/upcoming_changes/23789.new_feature.rst:1]:
```suggestion
```

Comment (mattip) [reply doc/release/upcoming_changes/23789.new_feature.rst:1]: The Pr to add more array API used this PR's number in its release note, causing a conflict. I will resolve it by removing the release note here, it is not really needed.

**Comment by mattip** [comment]
Thanks @bwalshe  and @asmeurer

File: doc/release/upcoming_changes/23789.new_feature.rst
Comment (asmeurer) [reply doc/release/upcoming_changes/23789.new_feature.rst]: Ooops I must have copied the wrong number. BTW have you considered just getting the number automatically instead of making people put it in the file name?

Comment (mattip) [reply doc/release/upcoming_changes/23789.new_feature.rst]: I think this is a shortcoming of the towncrier tool we use. The number is used to create a link in the release docs. We could move to a different tool, [scriv](https://scriv.readthedocs.io/en/1.1.0/philosophy.html) comes to mind.

**Comment by bwalshe** [comment]
Hi again. I was wondering - should I have cleaned up my commit history at some point before this was merged, or does that matter?

**Pull request merged**
